src/state/tls.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator resource definition."""
+
+import dataclasses
+from typing import Dict
+
+import ops
+from ops.jujuversion import JujuVersion
+from ops.model import Relation
+
+TLS_CERTIFICATES_INTEGRATION = "certificates"
+
+
+class TlsIntegrationMissingError(Exception):
+    """Exception raised when _situation_.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the TlsIntegrationMissingError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+@dataclasses.dataclass(frozen=True)
+class TLSInformation:
+    """A component of charm state containing resource definition for kubernetes secret.
+
+    Attrs:
+        tls_requirer_integration: The integration instance with a TLS provider.
+        tls_certs: A dict of hostname: certificate obtained from the relation.
+        tls_keys: A dict of hostname: private_key stored in juju secrets.
+    """
+
+    tls_requirer_integration: Relation
+    tls_certs: Dict[str, str]
+    tls_keys: Dict[str, str]
+
+    @classmethod
+    def from_charm(cls, charm: ops.CharmBase) -> "TLSInformation":
+        """Create a resource definition from charm instance.
+
+        Args:
+            charm (ops.CharmBase): The gateway-api-integrator charm.
+
+        Raises:
+            TlsIntegrationMissingError: When integration is not ready.
+
+        Returns:
+            TLSInformation: Information about configured TLS certs.
+        """
+        tls_requirer_integration = charm.model.get_relation(TLS_CERTIFICATES_INTEGRATION)
```

Comment:
I tried things out a bit and the issue here is that when we store the relation object here (in charm state), the validation ensures that the relation exists and the relation object in charm state is of type `Relation` (rather than `Relation | Any` when doing `model.get_relation()`). Since the lib and our `tls_relation` module are designed in different ways (the lib doesn't assume that the relation always exist while our `tls_relation` module does the opposite and assumes that the relation exist). We'd have to significantly rewrite one or the other for them to work. I think we'd need to discuss on this further and if changes are required a follow-up PR would be more suited. I'm marking this as resolved for now. 

Thanks!